### PR TITLE
fix(features): five feature-service findings from the 2026-08-30 audit

### PR DESCRIPTION
**Pull request by ishiland**
Five findings from the codebase audit 2026-08-30 (8dc529f17), tracked in #1778. All five still reproduced on current main (base cac6842b7); none had been fixed by an earlier merge. One commit per finding, each with a test that fails on main.

## Dataset rows endpoint interpolates bare column identifiers, so a reserved-word column permanently 500s the table browser

`get_dataset_rows` interpolated column names bare into the SELECT list and the ILIKE filter clause, so `SELECT gid, desc FROM data.x` was a 42601 syntax error. 42601 is in none of the sqlstate sets that function degrades on, so the endpoint returned 5xx for that dataset forever, for anonymous callers on a public dataset too. Nothing renames SQL keywords on ingest: `rename_reserved_columns` acts only on gid/geom/geometry/geom_4326/fid/ogc_fid, while `desc`, `order`, `user`, `group`, `end` and `to` are routine ogr2ogr output from DBF fields.

Both sites now emit through `_safe_column_ref`, a helper in the dataset domain's `_sql_safety` module that mirrors the escaping already used by `layers.service._qcol`, `features.service.live_property_columns` and `processing.ingest.metadata_sql._sql_quote_ident`. Membership tests still compare bare names, so quoting happens at emission.

Counterfactual: `tests/test_dataset_rows_reserved_columns.py` (2 tests) both fail on main with `sqlalchemy.exc.ProgrammingError` 42601.

## OGC Features property filter on any non-text column returns 503, even with a valid value

Property filters bound the raw query-string value, so SQLAlchemy typed the bind VARCHAR and PostgreSQL has no `bigint = character varying` operator. Every non-text property filter failed with 42883, on both the OGC items endpoint and the native features list, and the OGC handler's caller-fault branch was gated on a CQL2 `filter=` being present, so it fell through to `503 Dataset table is temporarily unavailable`. Clients retried forever against a query-shape bug and the queryables document led conformant clients straight into it.

Each filter value is now bound with the database type of the column it names, read from the live schema, covering exactly the types the queryables document advertises as filterable. The numeric families keep their own database types for the reason recorded in `standards/ogc/filtering.py`: a float8-cast bind against a REAL column promotes the stored float4 and stops comparing equal. A value that does not parse is a 400 naming the property, decided before any database round trip. A column of an unadvertised type keeps the raw string bind, and both routers now classify the type-shaped sqlstate that produces as the caller's 400.

Counterfactual: `tests/test_features_property_filter_types.py`, 11 of 13 fail on main (the two text-column cases pass, which is why the bug survived CI).

## Feature pagination runs a full filtered COUNT(*) on every page whenever any filter is active

The cached feature_count fast path applies only to a completely unfiltered request, and the count deliberately strips the keyset cursor, so one bbox or property filter put a full filtered COUNT(*) on every page, including the keyset pages whose entire purpose is constant-time access.

I took the cap-with-an-estimate-flag option rather than compute-once-per-cursor-chain, because it needs no new query parameter and therefore no OpenAPI or SDK change. The count now runs inside a LIMIT: exact up to 20000 matches, and past that the planner's row estimate answers. The estimate is never reported below the rows already counted, so a `next` link keyed on `offset + limit < total` cannot truncate pagination at the cap.

`numberMatched` keeps its place in both response schemas. Which of the two a page carries rides on a new `X-GeoLens-Number-Matched: estimated` response header, present only when estimated, added to both CORS expose lists so a browser client can read it. Response headers are not part of the OpenAPI snapshot, so `make openapi-check` stays clean.

Counterfactual: `tests/test_features_bounded_count.py` (7 tests). On main `get_features` returns two values and never bounds the count, so the estimate cases fail.

## Every single-feature edit runs a full-table COUNT(*)+ST_Extent seq scan

`_refresh_count_and_extent` runs one unqualified aggregate over the whole table after every insert, replace, delete and geometry-bearing patch, plus a second scan over `ST_ShiftLongitude` past 180 degrees of width and a third `SELECT DISTINCT GeometryType(...)` for created datasets. There is no bulk feature endpoint, so digitizing 200 points costs 200 of them.

The write handlers now say how many rows the write added or removed and where the geometry it touched sits: the new envelope from the request body, the prior envelope from a primary-key lookup before the row is moved or deleted. When every touched envelope is strictly inside the stored extent, the extent provably did not change and no aggregate runs; feature_count moves by the delta, as a SQL-side expression so two concurrent writes cannot both read N and write N+1.

Three guards keep that sound, each documented at the code. The containment test is strict, because a row that only touches a side may be the row defining it. A two-ring antimeridian extent is refused outright: ST_XMin/ST_XMax of the MULTIPOLYGON `seam_extent_wkt_for_table` produces are -180/180, so a longitude in the gap would test as inside a box the dataset never occupies. And on a created layer with a generic geometry column the fast path covers inserts only, because a removed or moved geometry can narrow the derived display type.

Counterfactual: `tests/test_features_incremental_metadata.py` (22 tests), 16 fail on main.

## A property whose column name fails the write path's stricter regex is silently dropped with a 200/201

`_reject_unknown_properties` admitted any key present in column_info and the write loops then skipped whatever `_COLUMN_NAME_RE` rejected, a strictly narrower pattern than every producer of column_info. A dataset with a `_notes` column returned that key from GET but could not write it: POST and PUT answered 201/200 with the value never stored, and PUT did not even NULL the column, contradicting its own documented replacement semantics.

A supplied property naming a column the write path cannot address is now a 422 naming the column. PUT is stricter: replace writes every known column, so one unwritable column in column_info makes the documented replacement impossible whether or not the request mentions it. The hole is also closed where it opens, per the audit's note about the 63-character bound: `POST /datasets/create/` refuses a column name the write path could not address, so no such column gets built. That covers the over-63-character case, where PostgreSQL truncates the identifier in the DDL while column_info keeps the full string. The bound lands in the service rather than on `ColumnDefinition.name`, which keeps the OpenAPI snapshot unchanged.

Counterfactual: `tests/test_features_unwritable_columns.py` (8 tests), 5 fail on main.

## Schema, API and config impact

No request or response schema changed. `backend/openapi.json` is unchanged (`dump_openapi.py --check` exits 0), so no SDK, CLI or frontend-types regeneration. 422 was already declared in `ERROR_RESPONSES_WRITE`. The one API-surface addition is the `X-GeoLens-Number-Matched` response header and its two entries in the CORS expose lists. No migration, no config.

## Verification

Run from this worktree with `.env.test` against Postgres on localhost:5434 (the file is gitignored and was deleted afterwards).

- `uv run pytest tests/test_dataset_rows_reserved_columns.py tests/test_features_property_filter_types.py tests/test_features_bounded_count.py tests/test_features_unwritable_columns.py tests/test_features_incremental_metadata.py tests/test_features_crud.py tests/test_layering.py tests/test_cors_range_headers_1540.py tests/test_cors_vary_origin_1602.py tests/test_search_cors_1596.py -q` -> 202 passed
- `uv run pytest tests/test_ogc_features.py tests/test_ogc_features_filter.py tests/test_ogc_pagination.py tests/test_ogc_queryables.py tests/test_ogc_public_access.py -q` -> 135 passed
- `uv run pytest tests/test_ogc_cql2_filtering.py tests/test_ogc_collection_metadata.py tests/test_dataset_rows.py tests/test_datasets.py tests/test_rule1_structural.py tests/test_standards_layering.py tests/test_dataset_rows_db_errors.py -q` -> 155 passed
- `uv run pytest tests/test_stac_api.py tests/test_stac_integration.py tests/test_stac_visibility.py tests/test_export.py tests/test_export_access.py tests/test_export_antimeridian.py tests/test_feature_geometry_limits.py tests/test_features_geojson_z.py -q` -> 152 passed
- `uv run pytest tests/test_analysis_materialize.py tests/test_analysis_preview.py tests/test_audit.py tests/test_dp01_ingest_write_path.py tests/test_sql_safety.py tests/test_codeql_qtable_suppressions.py tests/test_rule2_structural.py -q` -> 345 passed
- `uv run pytest tests/test_layering.py -q` -> 47 passed
- `uv run ruff check . && uv run ruff format --check .` -> clean, 1105 files formatted
- `PYTHONPATH=. uv run python scripts/dump_openapi.py --check` -> exit 0

Each counterfactual was measured by reverting only `backend/app/` to the commit's parent, rerunning the new test file, and restoring.

## Caps raised

- `backend/app/modules/catalog/datasets/domain/service_query.py`: 429 -> 439
- `backend/app/modules/catalog/features/service.py`: first entry, 1055, then 1135, 1187, 1384 as the later commits landed. It crossed the 1000-line inclusion threshold in this PR. The entry records the clean split for when it next grows: everything below the "Write operations" banner moves to a sibling module behind a re-export facade, because `standards/ogc/router.py` and `standards/stac/router.py` may import `features.service` and nothing else under features, and several tests reach private names through it.

## Left alone

- `features/router.py` optimistic concurrency, out of scope for this lane.
- The antimeridian seam trigger still fires on naive extent width rather than an actual crossing, so a globe-spanning dataset pays the shifted-domain scan whenever it does recompute. Deciding a real crossing needs the shifted aggregate the check exists to avoid, and the insert and edit fast paths take it off the common path.
- The audit also suggested a `SET LOCAL statement_timeout` around the inline recompute. Not done: a statement timeout aborts the surrounding transaction, so it would fail the user's edit rather than degrade the metadata, unless the aggregate is moved into its own savepoint.
- The audit's optional suggestion of adding 42601 to a sqlstate set was not taken. With the identifiers quoted, that path cannot emit 42601, and a future escape is better caught by the quoting than reclassified.

## Review round 1

Both findings were in `features/service.py`, fixed in 7b4bf0855.

**P1, pagination depending on the estimated total.** With the exact count capped at 20000 and a planner underestimate, `total` clamps to 20001, and the native features router decided `next` from `offset + limit < total`: at offset 20000 with limit 200 a full page came back with no link to the remaining rows. `get_features` now fetches one row past the page, drops it from the payload, and reports `has_more` on a `FeaturePage`; the native handler and the OGC items handler both build `next` from that. The over-fetch the OGC handler used to do for itself moved into the service, so `limit + 1` happens once rather than in each caller. STAC shares no pagination here: `standards/stac/router.py` imports only `parse_bbox` from this module. `total` is additionally floored at the rows already served plus the one more known to exist, so `numberMatched` cannot contradict the features beside it.

**P2, the prior-bounds read racing the mutation.** The envelope a write overwrites is now captured by the mutating statement: `DELETE ... RETURNING` for deletes, and for the two update paths a CTE that takes `FOR UPDATE` on the target row and returns the values it overwrote, so the envelope is the version that statement actually replaced. `replace_feature` and `update_feature` return a `FeatureWrite(feature, prior_bounds)` and the routers read it from the write; the unlocked helper is deleted rather than left for a future caller.

The same push closes the half the prior-bounds capture did not: the stored extent was also read unlocked, so two writers on one dataset could still interleave read-decide-write, one skipping the recompute because its geometry was inside the extent it read while the other shrank that extent from an aggregate taken before the first row landed. Both metadata paths now read the extent `FOR UPDATE` before either decides anything. Every writer that reaches `refresh_dataset_metadata` updates the dataset row before commit anyway, so it already serialized with its peers; the lock only moves that serialization ahead of the decision that depends on it, and both paths take the record row before the dataset row, matching the ORM's flush order.

### Round 1 verification

- `uv run pytest tests/test_features_bounded_count.py tests/test_features_incremental_metadata.py tests/test_features_unwritable_columns.py tests/test_features_property_filter_types.py tests/test_features_crud.py tests/test_features_geojson_z.py tests/test_export_antimeridian.py tests/test_extent_producers_934.py tests/test_dataset_rows_reserved_columns.py -q` -> 152 passed
- `uv run pytest tests/test_ogc_features.py tests/test_ogc_features_filter.py tests/test_ogc_pagination.py tests/test_ogc_queryables.py tests/test_ogc_public_access.py -q` -> 135 passed
- `uv run pytest tests/test_stac_api.py tests/test_stac_integration.py tests/test_export.py tests/test_export_access.py tests/test_rule1_structural.py tests/test_standards_layering.py tests/test_audit.py tests/test_analysis_preview.py -q` -> 203 passed
- `uv run pytest tests/test_layering.py -q` -> 47 passed
- `uv run ruff check . && uv run ruff format --check .` -> clean
- `PYTHONPATH=. uv run python scripts/dump_openapi.py --check` -> exit 0

New pins. P1: 20300 matching rows with `_planner_row_estimate` forced to 0, offset 20000 limit 200 emits `next` with `offset=20200` and `numberMatched >= 20200`; the last page at offset 20200 returns 100 rows and no `next`; the same boundary through the OGC items handler; and a service-level assertion that `has_more` is true while `total_is_estimate` is true. P2: two sessions on :5434, where session A reads the envelope with a plain SELECT, a second session moves the feature far outside the extent and commits, and A then deletes or replaces it; the captured envelope equals the moved-to point, differs from A's own earlier read, and the refresh runs the aggregate rather than the fast path. Plus a NULL-geometry delete reporting no bounds, and missing-gid raises on both paths.

Counterfactual against the parent commit: 9 of these fail, the P1 link case with "a full page with rows remaining must link on".

Cap for `features/service.py` raised 1384 -> 1491 in the same commit.

## Review round 2

One P2 on the native handler's exception tuple, fixed in 6f36932a0. Measuring it first changed what the fix had to be, so both corrections are recorded here.

The reported example does not reproduce as a 500. `?ratio=1e100` on a REAL column goes through the property-filter path, where SQLAlchemy's `REAL()` bind emits no narrowing cast: PostgreSQL promotes the stored float4, compares in float8, and answers 200 with zero features. `?construction_year=2147483648` on an INTEGER column is the same shape, bound BigInteger, where `int4 = int8` is a legal comparison no stored value can satisfy. Both are worse than the reported 500, because a confident wrong answer is harder to notice than an error.

SQLSTATE 22003 is real, on the other path: `filter=ratio = 1e100` compiles to `ratio = CAST($1::FLOAT AS REAL)` and that cast does overflow. The exception class is not `DataError` either. SQLAlchemy's asyncpg dialect wraps both `NumericValueOutOfRangeError` and `asyncpg.exceptions.DataError` into the base `sqlalchemy.exc.DBAPIError`, so `except DataError` never fired and the OGC handler 500'd on it too; adding `DataError` to the native tuple would have fixed nothing.

Both routers now catch `DBAPIError` and classify through one shared `is_caller_type_fault` in `core/db/sqlstate.py`. They had genuinely drifted, which is the half of the finding that held exactly: the OGC handler carried the type-fault set inline while the native list tested a narrower `BAD_QUERY_INPUT`, so one failure was a 400 through one endpoint and a 503 through the other. The 503 and 500 branches keep their existing shape, so an unrecognized state stays an honest 500 rather than being reported as an outage.

The pre-execution range checks are what actually turn the two pinned cases into 400s. `core/db/pg_ranges.py` holds the int2/int4/int8 bounds and the float4 magnitude; the property-filter path and the CQL2 comparison path both read it, which also replaces the single int8 bound CQL2 had been applying to smallint, integer and bigint alike.

Every typed bind built from a caller value in the features service, and where each now goes:

| bind | handling |
| --- | --- |
| property filters | parse, then `check_pg_value_range`, a 400 naming the property and the bound |
| CQL2 comparison literals | the same table, through `_checked_value` |
| limit / offset / after_gid | `check_int8_range`, a 400 naming the parameter. FastAPI's `int` has no upper bound, so `?offset=10**23` was a 500 on both routers |
| bbox | binds float8, which has no narrowing step to overflow, and `parse_bbox` already refuses non-finite values. Left alone |
| OGC `datetime` | reaches the pagination-link builder only, never a bind |

### Round 2 verification

- `uv run pytest tests/test_features_property_filter_types.py -q` -> 28 passed
- `uv run pytest tests/test_ogc_features.py tests/test_ogc_features_filter.py tests/test_ogc_cql2_filtering.py tests/test_ogc_queryables.py tests/test_ogc_pagination.py tests/test_ogc_public_access.py -q` -> 165 passed
- `uv run pytest tests/test_features_bounded_count.py tests/test_features_incremental_metadata.py tests/test_features_unwritable_columns.py tests/test_features_crud.py tests/test_features_geojson_z.py tests/test_dataset_rows_db_errors.py tests/test_dataset_rows_reserved_columns.py tests/test_export_antimeridian.py -q` -> 152 passed
- `uv run pytest tests/test_stac_api.py tests/test_stac_integration.py tests/test_stac_search_validation.py tests/test_export.py tests/test_export_access.py tests/test_rule1_structural.py tests/test_standards_layering.py tests/test_datasets.py -q` -> 175 passed
- `uv run pytest tests/test_layering.py -q` -> 47 passed
- `uv run ruff check . && uv run ruff format --check .` -> clean
- `PYTHONPATH=. uv run python scripts/dump_openapi.py --check` -> exit 0

New pins: `1e100` on REAL and `2147483648` on INTEGER are 400s naming the property and the bound on both routers; the same two through CQL2; `offset` and `after_gid` past int8 on both; plus vacuity guards that in-range values still match, that `offset = 2**63 - 1` still answers 200, and that a huge finite bbox is not a 500. Against the parent commit 9 of these fail, 5 as silent 200s and 4 as uncaught 500s.

Cap for `features/service.py` raised 1491 -> 1507 in the same commit.

## Review round 3

One P2 on the total floor added in round 1, fixed in 917c37005. Reproduced against the parent commit before changing anything: five matches asked for at offset 100 reported `numberMatched: 100` on both routers, and a keyset page with `after_gid` plus `offset=100` reported 111 against a thirty-row match set.

The r1 floor read `offset + len(rows)` unconditionally, so it invented matches out of the offset rather than reporting what the query found. It is now a claim about rows that exist, and only ever raises an estimate:

- An exact count is never raised. It already counted the whole match set, so anything above it is fiction. That also stops a full page of an exact count reporting `count + 1`.
- An empty page proves nothing and leaves the total alone.
- A keyset page floors by the rows in hand only, because the query ignores `offset` when `after_gid` is set and borrowing an offset it never applied would invent matches the same way.

The property the floor exists for is unchanged: an estimate below the rows a page is about to show still floors to `offset + len(rows) + 1`, so `numberMatched` cannot contradict the features beside it in the same response.

Extracted to `_floor_estimated_total` rather than narrowed in place, because `get_features` was back at ruff's C901 ceiling at 16 and because each of the three conditions is a separate claim that wants its reason recorded beside it.

### Round 3 verification

- `uv run pytest tests/test_layering.py tests/test_features_bounded_count.py -q` -> 64 passed
- `uv run pytest tests/test_ogc_features.py tests/test_ogc_features_filter.py tests/test_ogc_pagination.py tests/test_ogc_cql2_filtering.py tests/test_ogc_queryables.py tests/test_ogc_public_access.py -q` -> 165 passed
- `uv run pytest tests/test_features_property_filter_types.py tests/test_features_incremental_metadata.py tests/test_features_unwritable_columns.py tests/test_features_crud.py tests/test_features_geojson_z.py tests/test_export_antimeridian.py tests/test_dataset_rows_reserved_columns.py -q` -> 143 passed
- `uv run pytest tests/test_stac_api.py tests/test_stac_integration.py tests/test_export.py tests/test_export_access.py tests/test_rule1_structural.py tests/test_standards_layering.py tests/test_datasets.py -q` -> 167 passed
- `uv run ruff check . && uv run ruff format --check .` -> clean
- `PYTHONPATH=. uv run python scripts/dump_openapi.py --check` -> exit 0

New pins: five matches at offset 100 report 5 with no `next`, on the native endpoint and on OGC items; a keyset page with `after_gid` and a bogus `offset=100` reports the true 30; an exact count with a full page stays at the count rather than count + 1; an estimated low count with a full page still floors to exactly `20000 + 200 + 1`; and an estimated count on a page past the end is not floored at all. Four fail on the parent commit, with the measured values 100, 100, 111 and 21300.

Cap for `features/service.py` raised 1507 -> 1541 in the same commit.

### CODE REVIEW

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/features/service.py [hunk 1]
```diff
@@ -241,6 +411,123 @@ async def _projected_row_source(
...
+                f"SELECT COUNT(*) FROM (SELECT 1 FROM {quoted} t {where_sql} "
+                f"LIMIT :count_cap) s"
+            ).bindparams(**binds, count_cap=_FILTERED_COUNT_CAP + 1)
+        )
+    )
+    counted = int(capped_result.scalar_one())
+    if counted <= _FILTERED_COUNT_CAP:
+        return counted, False
+    estimate = await _planner_row_estimate(db, quoted, where_sql, binds, apply_binds)
+    return max(counted, estimate), True
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/features/service.py:549]: Fixed in 7b4bf0855.

Pagination no longer reads the count at all. `get_features` fetches one row past the page, drops it from the payload, and returns `has_more` on a `FeaturePage`; the native features router and the OGC items handler both build `next` from that. The over-fetch the OGC handler used to do for itself moved into the service, so `limit + 1` happens once and both callers get the same answer rather than deriving their own. STAC shares nothing here: `standards/stac/router.py` imports only `parse_bbox` from this module, so there is no third `next`-link builder to update.

`total` is also floored at `offset + len(rows)`, plus one more when `has_more`, so numberMatched can never claim fewer rows than the features sitting beside it in the same response.

Pinned in `tests/test_features_bounded_count.py`: 20300 matching rows, `_planner_row_estimate` forced to 0, offset 20000 limit 200 emits `next` with `offset=20200` and `numberMatched >= 20200`; the last page at offset 20200 returns 100 rows and no `next`. The same boundary is asserted through the OGC items handler, plus a service-level test that `has_more` is true while `total_is_estimate` is true. On the parent commit the first of those fails with "a full page with rows remaining must link on".

File: backend/app/modules/catalog/features/service.py [hunk 2]
```diff
@@ -850,12 +1152,209 @@ async def _derive_created_geometry_type(session: AsyncSession, table_name: str)
...
+
+    A primary-key lookup, so callers can afford it on every write to learn
+    whether the row they are about to move or delete could have been defining
+    the dataset's extent.
+    """
+    result = await db.execute(
+        text(
+            f"SELECT ST_XMin(geom_4326), ST_YMin(geom_4326), "
+            f"ST_XMax(geom_4326), ST_YMax(geom_4326) "
+            f"FROM {get_catalog_port().quote_table(table_name)} WHERE gid = :gid"
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/features/service.py]: Fixed in 7b4bf0855, using the RETURNING shape.

`delete_feature` now runs `DELETE ... WHERE gid = :gid RETURNING ST_XMin(geom_4326), ...` and returns the envelope it removed. The two UPDATE paths use `WITH prior AS (SELECT gid, ST_XMin(geom_4326) ... FROM t WHERE gid = :gid FOR UPDATE) UPDATE t AS t SET ... FROM prior WHERE t.gid = prior.gid RETURNING prior....`, so the envelope is the version that statement actually overwrote: the CTE waits on a concurrent writer and re-reads at the latest committed version. `replace_feature` and `update_feature` return a `FeatureWrite(feature, prior_bounds)` and the routers read the envelope from the write instead of taking one beforehand. The unlocked helper is deleted rather than left in place for a future caller to reach for.

One thing the prior-bounds capture alone did not close, so it is in the same push: the stored extent was also read in an unlocked statement, so two writers on one dataset could still interleave read-decide-write, one skipping the recompute because its geometry was inside the extent it read while the other shrank that extent from an aggregate taken before the first row landed. Both metadata paths now read the extent `FOR UPDATE` before either decides anything. Every writer that reaches `refresh_dataset_metadata` goes on to update the dataset row anyway, so it already serialized with its peers at commit; the lock only moves that serialization ahead of the decision that depends on it. Both paths take the record row before the dataset row, which is the order the ORM flushes them in.

Pinned in `tests/test_features_incremental_metadata.py` with two sessions on :5434. Session A reads the feature's envelope with a plain SELECT, a second session moves it far outside the extent and commits, then A deletes (and, in the sibling test, replaces). The test asserts the captured envelope equals the moved-to point, that it differs from what A's own earlier read said, and that the refresh then runs the aggregate rather than the fast path. Two more tests pin that a deleted row with NULL geometry reports no bounds rather than a zero-size envelope, and that a missing gid still raises.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/features/router.py [hunk 3]
```diff
@@ -359,7 +362,26 @@ async def list_features(
...
         )
-    except (ProgrammingError, OperationalError):
+    except ValueError as e:
+        # fix(#1778): an unparseable property-filter value is the caller's bug,
+        # rejected before the query runs.
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail=str(e),
+        )
+    except (ProgrammingError, OperationalError) as exc:
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/features/router.py]: Fixed in 6f36932a0, though the measurement changed what the fix had to be. Two corrections to the premise first, both measured on PostgreSQL 18 with asyncpg rather than reasoned about.

`?ratio=1e100` on a REAL column is not a 500 and never reached 22003. SQLAlchemy's `REAL()` bind emits no narrowing cast on the property-filter path, so PostgreSQL promotes the stored float4 and compares in float8: HTTP 200 with zero features. `?construction_year=2147483648` on an INTEGER column is the same shape, bound BigInteger, and `int4 = int8` is a legal comparison no stored value can satisfy: 200 with zero features. Both are worse than the reported 500, because a confident wrong answer is harder to notice than an error.

The 22003 you predicted is real, just on the other path: `filter=ratio = 1e100` compiles to `ratio = CAST($1::FLOAT AS REAL)`, and that cast does overflow. And the exception class is not `DataError`. SQLAlchemy's asyncpg dialect wraps both `asyncpg.exceptions.NumericValueOutOfRangeError` and `asyncpg.exceptions.DataError` into the base `sqlalchemy.exc.DBAPIError`, so `except DataError` never fired for either and the OGC handler 500'd on it too. Adding `DataError` to the native tuple would not have fixed anything.

So: both routers now catch `DBAPIError` and classify through one shared `is_caller_type_fault` in `core/db/sqlstate.py`. They had genuinely drifted, which is the half of your finding that held exactly: the OGC handler carried the type-fault set inline while the native list tested a narrower `BAD_QUERY_INPUT`. The 503 and 500 branches keep their shape, so an unrecognized state stays an honest 500 rather than being reported as an outage.

And the pre-execution range checks you suggested, which are what actually make the two pinned cases 400s. `core/db/pg_ranges.py` holds int2/int4/int8 and the float4 bound; the property-filter path and the CQL2 comparison path both read it, so CQL2's single int8 bound for smallint, integer and bigint alike is gone.

Enumeration of every typed bind built from a caller value in the features service:

| bind | handling |
| --- | --- |
| property filters | parse, then `check_pg_value_range` -> 400 naming property and bound |
| CQL2 comparison literals | same table, via `_checked_value` |
| limit / offset / after_gid | `check_int8_range` -> 400 naming the parameter. FastAPI's `int` has no upper bound, so `?offset=10**23` was a 500 on both routers |
| bbox | binds float8, no narrowing step to overflow; `parse_bbox` already refuses non-finite. Left alone, reason recorded |
| OGC `datetime` | reaches the link builder only, never a bind |

Pins in `tests/test_features_property_filter_types.py`: `1e100` on REAL and `2147483648` on INTEGER are 400s naming the property and the bound, on both routers; the same two through CQL2; `offset` and `after_gid` past int8 on both; plus vacuity guards that in-range values still match, that `offset = 2**63 - 1` still answers 200, and that a huge finite bbox is not a 500. Against the parent commit 9 of these fail: 5 as silent 200s and 4 as uncaught 500s.

Cap for `features/service.py` 1491 -> 1507.

**Comment by ishiland** [comment]
@codex review

File: backend/app/modules/catalog/features/service.py [hunk 4]
```diff
@@ -370,23 +719,24 @@ def _with_cql2_binds(stmt):
...
         )
-        total = count_result.scalar_one()
 
-    return rows, total
+    # fix(#1778 review r1): the total never claims fewer rows than this page has
+    # already shown, plus the one more we know is there. An estimate below the
+    # cap, or a stale cached feature_count, would otherwise make numberMatched
+    # contradict the features beside it in the same response.
+    served = offset + len(rows)
+    total = max(total, served + 1 if has_more else served)
```
[the remark this answers is not in the document]
Comment (ishiland) [reply backend/app/modules/catalog/features/service.py]: Fixed in 917c37005. Confirmed against the parent commit first: five matches at offset 100 reported `numberMatched: 100` on both routers, and a keyset page with `after_gid` plus `offset=100` reported 111 against a thirty-row match set. Both reproduce exactly as described.

The floor is now a claim about rows that exist, extracted to `_floor_estimated_total` with the three conditions you named:

- An exact count is never raised. It already counted the whole match set, so anything above it is fiction. This subsumes the empty-page case for exact counts and also stops a full page of an exact count reporting `count + 1`.
- An empty page proves nothing, so it leaves the total alone whether or not the count is an estimate.
- A keyset page passes `offset=0` into the floor, because the query ignores `offset` when `after_gid` is set. Borrowing an offset its own query never applied would invent matches the same way. The rows in hand are all such a page can prove.

The r1 property it exists for is unchanged: an estimate below the rows a page is about to show still floors to `offset + len(rows) + 1`.

Extracted rather than narrowed in place because `get_features` was back at ruff's C901 ceiling at 16, and because each condition is a separate claim that wants its reason recorded beside it.

Pins in `tests/test_features_bounded_count.py`: five matches at offset 100 report 5 with no `next`, on the native endpoint and on OGC items; a keyset page with `after_gid` and a bogus `offset=100` reports the true 30; an exact count with a full page stays at the count rather than count + 1; an estimated low count with a full page still floors to exactly `20000 + 200 + 1`; and an estimated count on a page past the end is not floored at all. Four of those fail on the parent commit, with the measured values 100, 100, 111, and 21300.

Cap for `features/service.py` 1507 -> 1541.

**Comment by ishiland** [comment]
@codex review
